DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs

```diff
@@ -0,0 +1,304 @@
+using DragaliaAPI.Features.Fort;
+using DragaliaAPI.Shared.Definitions.Enums;
+using DragaliaAPI.Shared.Features.StorySkip;
+using DragaliaAPI.Shared.MasterAsset;
+using DragaliaAPI.Shared.MasterAsset.Models;
+using DragaliaAPI.Shared.MasterAsset.Models.Story;
+using static DragaliaAPI.Shared.Features.StorySkip.StorySkipRewards;
+
+namespace DragaliaAPI.Features.StorySkip;
+
+public class StorySkipService(
+    ApiContext apiContext,
+    IFortRepository fortRepository,
+    ILogger<StorySkipService> logger,
+    IQuestRepository questRepository,
+    IStoryRepository storyRepository,
+    IUserDataRepository userDataRepository
+)
+{
+    public void IncreaseFortLevels(long viewerId)
+    {
+        Dictionary<FortPlants, FortConfig> fortConfigs = StorySkipRewards.FortConfigs;
+        IQueryable<DbFortBuild> userForts = fortRepository.Builds.Where(x =>
+            x.ViewerId == viewerId
+        );
+        List<DbFortBuild> newUserForts = new();
+
+        foreach ((FortPlants fortPlant, FortConfig fortConfig) in fortConfigs)
+        {
+            IQueryable<DbFortBuild> fortsToUpdate = userForts.Where(x => x.PlantId == fortPlant);
+
+            foreach (DbFortBuild fortToUpdate in fortsToUpdate)
+            {
+                if (fortToUpdate.Level < fortConfig.Level)
+                {
+                    logger.LogDebug("Updating fort at BuildId {buildId}", fortToUpdate.BuildId);
+                    fortToUpdate.Level = fortConfig.Level;
+                    fortToUpdate.BuildStartDate = DateTimeOffset.UnixEpoch;
+                    fortToUpdate.BuildEndDate = DateTimeOffset.UnixEpoch;
+                }
+            }
+
+            for (int x = fortsToUpdate.Count(); x < fortConfig.BuildCount; x++)
+            {
+                logger.LogDebug("Adding fort {plantId}", fortPlant);
+                DbFortBuild newUserFort =
+                    new()
+                    {
+                        ViewerId = viewerId,
+                        PlantId = fortPlant,
+                        Level = fortConfig.Level,
+                        PositionX = fortConfig.PositionX,
+                        PositionZ = fortConfig.PositionZ,
+                        BuildStartDate = DateTimeOffset.UnixEpoch,
+                        BuildEndDate = DateTimeOffset.UnixEpoch,
+                        IsNew = true,
+                        LastIncomeDate = DateTimeOffset.UnixEpoch
+                    };
+                newUserForts.Add(newUserFort);
+            }
+        }
+
+        if (newUserForts.Count > 0)
+        {
+            apiContext.PlayerFortBuilds.AddRange(newUserForts);
+        }
+    }
+
+    public int ProcessQuestCompletions(long viewerId)
+    {
+        int wyrmite = 0;
+        FrozenSet<QuestData> questDatas = MasterAsset
+            .QuestData.Enumerable.Where(x =>
+                x.Gid < 10011 && x.Id > 100000000 && x.Id.ToString().Substring(6, 1) == "1"
+            )
+            .ToFrozenSet();
```

Comment:
This will still get computed on every execution of this method. You need to hoist it as a `static readonly` variable in the class itself so that it is shared between instances:

```cs
public class StorySkipService
{
    private static readonly FrozenSet<QuestData> // ...
```